Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm

```diff
@@ -136,30 +154,80 @@ static bool isNonRecoverableError(OSStatus status)
+    if (auto status = PAL::CMVideoFormatDescriptionCreateForImageBuffer(kCFAllocatorDefault, imageBuffer, &rawImageBufferDescription); status != noErr)
         return makeUnexpected(status);
-
     RetainPtr<CMVideoFormatDescriptionRef> imageBufferDescription = adoptCF(rawImageBufferDescription);
 
+#if PLATFORM(VISION)
+    if (s_canCopyFormatDescriptionExtension) {
+        static CFStringRef keys[] = {
+            PAL::kCMFormatDescriptionExtension_CameraCalibrationDataLensCollection,
+            PAL::kCMFormatDescriptionExtension_HasLeftStereoEyeView,
+            PAL::kCMFormatDescriptionExtension_HasRightStereoEyeView,
+            PAL::kCMFormatDescriptionExtension_HeroEye,
+            PAL::kCMFormatDescriptionExtension_HorizontalFieldOfView,
+            PAL::kCMFormatDescriptionExtension_HorizontalDisparityAdjustment,
+            PAL::kCMFormatDescriptionExtension_StereoCameraBaseline,
+            PAL::kCMFormatDescriptionExtension_ProjectionKind,
+            PAL::kCMFormatDescriptionExtension_ViewPackingKind
+        };
+        static constexpr size_t numberOfKeys = sizeof(keys) / sizeof(keys[0]);
+        size_t keysSet = 0;
+        Vector<RetainPtr<CFPropertyListRef>, numberOfKeys> values(numberOfKeys, [&](size_t index) -> RetainPtr<CFPropertyListRef> {
+            WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
+            RetainPtr value = PAL::CMFormatDescriptionGetExtension(description, keys[index]);
+            WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
+            if (!value)
+                return nullptr;
+            keysSet++;
+            return value;
+        });
+
+        if (keysSet) {
+            RetainPtr<CFMutableDictionaryRef> copyExtensions;
+            if (RetainPtr extensions = PAL::CMFormatDescriptionGetExtensions(rawImageBufferDescription))
+                copyExtensions = adoptCF(CFDictionaryCreateMutableCopy(kCFAllocatorDefault, CFDictionaryGetCount(extensions.get()) + keysSet, extensions.get()));
+            else
+                copyExtensions = adoptCF(CFDictionaryCreateMutable(kCFAllocatorDefault, keysSet, &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks));
+
+            for (size_t index = 0; index < numberOfKeys; index++) {
+                WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
+                if (values[index])
+                    CFDictionarySetValue(copyExtensions.get(), keys[index], values[index].get());
+                WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
+            }
+
+            auto codecType = CVPixelBufferGetPixelFormatType(imageBuffer);
+            auto width = CVPixelBufferGetWidth(imageBuffer);
+            auto height = CVPixelBufferGetHeight(imageBuffer);
+            if (auto status = PAL::CMVideoFormatDescriptionCreate(kCFAllocatorDefault, codecType, width, height, copyExtensions.get(), &rawImageBufferDescription); status != noErr)
+                return makeUnexpected(status);
+            imageBufferDescription = adoptCF(rawImageBufferDescription);
```

Comment:
Can we move this code to a separate function with a name like `copyDescriptionExtensionValues`?